index.js

```diff
@@ -3,57 +3,77 @@
 const fp = require('fastify-plugin')
 const formatEntry = require('./lib/formatEntry')
 const CRLF = '\r\n'
+const EarlyHints = `HTTP/1.1 103 Early Hint${CRLF}`
 
 function fastifyEarlyHints (fastify, opts, next) {
   if (fastify.initialConfig.http2 === true) {
-    return next(new Error('Early Hints cannot be used with a HTTP2 server.'))
+    return next(Error('Early Hints cannot be used with a HTTP2 server.'))
   }
 
   const formatEntryOpts = {
     warn: opts.warn
   }
 
-  function earlyHints (reply) {
-    const promiseBuffer = []
-    const serialize = function (c) {
-      let message = `HTTP/1.1 103 Early Hints${CRLF}`
-      for (let i = 0; i < c.length; i++) {
-        message += `${formatEntry(c[i], formatEntryOpts)}${CRLF}`
-      }
-      return `${message}${CRLF}`
-    }
-
-    return {
-      close: async function () {
-        if (promiseBuffer.length) {
-          await Promise.all(promiseBuffer)
+  fastify.decorateReply('writeEarlyHints', function (headers) {
+    const reply = this
+    let message = ''
+    if (Array.isArray(headers)) {
+      for (const nameValues of headers) {
+        if (typeof nameValues === 'object' && typeof nameValues.name === 'string' && typeof nameValues.value === 'string') {
+          message += `${nameValues.name}: ${nameValues.value}${CRLF}`
+        } else {
+          throw Error('"headers" expected to be name-value object')
```

Comment:
```suggestion
          return Promise.reject(Error('"headers" expected to be name-value object'))
```

otherwise the function sometimes returns a promise and sometimes not